src/ripple/basics/Number.h

```diff
@@ -387,6 +387,71 @@ class saveNumberRoundMode
     operator=(saveNumberRoundMode const&) = delete;
 };
 
+// Functors to facilitate intermediate steps rounding mode
+// in complex arithmetic expressions.
+// Class constructor saves the current rounding mode,
+// sets the requested rounding mode, and is called before
+// the arithmetic expression is evaluated. The rounding mode
+// state is restored in the functor call, which sets the outer
+// functor rounding mode or the initial mode. For instance, if
+// the following expression has to be minimized:
+// (a * b) / (c - d) then the functors are called as follows:
+// downard()(downward()(a * b) / upward()(c - d))
+// the call sequence is as follows:
+// set downward; set downward; num = a * b; restore downward; return n;
+// set upward; den = c - d; restore downward; return den; r = num / den;
+// restore to_nearest; return r;
+
+class upward
+{
+private:
+    Number::rounding_mode mode_;
+
+public:
+    upward() : mode_(Number::getround())
+    {
+        Number::setround(Number::rounding_mode::upward);
+    }
+    ~upward() = default;
+    upward(upward const&) = delete;
+    upward(upward const&&) = delete;
+    upward&
+    operator=(upward const&) = delete;
+    upward&
+    operator=(upward const&&) = delete;
+    Number const&
+    operator()(Number const& n)
+    {
+        Number::setround(mode_);
+        return n;
+    }
+};
+
+class downward
+{
+private:
+    Number::rounding_mode mode_;
+
+public:
+    downward() : mode_(Number::getround())
+    {
+        Number::setround(Number::rounding_mode::downward);
+    }
+    ~downward() = default;
+    downward(downward const&) = delete;
+    downward(downward const&&) = delete;
+    downward&
+    operator=(downward const&) = delete;
+    downward&
+    operator=(downward const&&) = delete;
+    Number const&
+    operator()(Number const& n)
+    {
+        Number::setround(mode_);
+        return n;
+    }
+};
```

Comment:
These are clever.  And I like the way they clean up the code.  But @scottschurr pointed out to me a potential to cause a run-time error:  Subexpressions in C++ are not guaranteed to be executed in any given order.  See: https://en.cppreference.com/w/cpp/language/eval_order#:~:text=Order%20of%20evaluation%20of%20any,same%20expression%20is%20evaluated%20again.

For example in:
```c++
z = downward()(a * b) / upward()(c - d);
```
I am concerned that the order could be:
1. construct downward
2. construct upward
3. execute a * b
4. execute c - d
5. call the downward() with a*b
6. call the upward with c-d
7. do the division

When I tried to simulate such an order with clang, I failed to do so.  But even if all of our compilers don't create a run-time error now, if they started to in the future, the error would be silent, and extremely subtle, hard to detect and hard to find and fix.

Reply:
Thanks for investigating. So have to break expressions into subexpressions. The functor would still work to make it less verbose, right? I.e.

    num = downard()(a * b);
    den = upward()(c - d);
    res = downard() (num / den);

Reply:
Yes, that will work.  We will have to watch out for maintenance in the future to make sure it doesn't combine such expressions.

For other's, here's what the above looks like without these helpers:
```c++
saveNumberRoundMode _{Number::setround(Number::downward)};
num = a * b;
Number::setround(Number::upward);
den = c - d;
Number::setround(Number::downward);
res = num / den;
```

The basic rule for both of these styles is that we need a `;` between calls to `Number::setround`.

Reply:
I'm wondering if parenthesis are going to force the order, for instance. Also note additional `downward()` to set the division rounding.

    z = downward()( (downward()(a * b) ) / ( upward()(c - d) ) )

Reply:
An approach that might allow combining expressions would be named arithmetic functions that take three parameters: two `Number's and a rounding mode.  So something vaguely like:
```
Number z = div(mul(a, b, downward), sub(c, d, upward), downward);
```
That's pretty ugly to read.  I haven't tried it, but it _might_ be a livable compromise?

Reply:
It is really hard to read. I think I'd rather break into subexpression if there is no better way. 
The standard does say this `Any expression in parentheses is also classified as a primary expression: this guarantees that the parentheses have higher precedence than any operator. Parentheses preserve value, type, and value category.` [standard](https://en.cppreference.com/w/cpp/language/expressions#:~:text=Any%20expression%20in%20parentheses%20is%20also%20classified%20as%20a%20primary%20expression) so everything in parenthesis should be evaluated first. So it should work as long as the constructor is called first.

Reply:
Regarding the constructor, here's the best I could find.  In the C++23 Standard it's section 6.7 Temporary objects [class.temporary] paragraph 2:

> The materialization of a temporary object is generally delayed as long as possible in order to avoid creating unnecessary temporary objects.

Note that it says "generally", which I take to mean it is not guaranteed.

Reply:
So the functor can't be used the way it's currently used. This works (I abbreviated dwn-downward, up-upward) and they are functions, no functors.

    Number z = dwn(dwn(a) + b)) / dwn(up(c) - d)

Can also add unary prefix functions to Number, which work like identity functions and set the rounding, for instance: `~` is downward and `&` is upward. Then:

    Number z = ~(~a + b) / ~(&c - d)

Or could use a variation of Scott's suggestion:
muld(a, b) -> mul(a, b, downward) and so on. Then:

    Number z = divd(muld(a, b), subu(c, d))

Reply:
It's a shame the functor syntax won't work, I liked that a lot.

Doesn't everything except:
```
Number z = div(mul(a, b, downward), sub(c, d, upward), downward);
```
Run into the order of operations problem? (Even if it doesn't I don't love the `dwn(dwn(a) + b)) / dwn(up(c) - d)` syntax). 

One thought: if we used expression templates we could use: 
```
z = downward(downward(a * b) / upward(c - d))
```
And get around the order of operations issue. However, expression templates aren't a lot of fun to write or debug. There are serious downsides to going that way. I'm probably opposed to doing that, but maybe it will spark an idea from someone else. (ref: expression templates: https://en.wikipedia.org/wiki/Expression_templates)

Reply:
Interesting. I glanced it over and need to take another pass but it gave me an idea. How about `upward` and `downward` classes with constructors to take Number and overloaded operators +,-,*,/ returning Number. Then it could be something like this. 

     z = downward(downward(a) * downward(b)) / downward((upward(c) - upward(d)))

If too verbose then can add an overload that takes Number; for inst. operator + (Number const&), then:

    z = downward(downward(a) * b)) / (upward(c) - d)

Reply:
I am seriously questioning if the cure is not worse than the disease.

There is also going to be some non-zero cost of setting the thread_local.  I'm not sure what that cost is.  It may be negligible.  But whatever it is, it is being doubled compared to doing this with calls to `Number::setround`.  If `Number::setround` is too verbose, perhaps we should just move `setround` outside of `Number`.

Reply:
It's not the problem with `setround`, it's breaking a complex arithmetic expression into subexpressions to get the rounding right. `Number` class made those expressions super easy to implement and read and the rounding makes it difficult.

Reply:
My suggestion is not to solve the "find a nice syntax for equations with rounding" in this PR. So something like we did in this patch: https://github.com/XRPLF/rippled/pull/5002/commits/b90f58d3a8ced5d7cbef12cd8f3215546f87ca25 and we can find a nice syntax later if we find the code in that style too difficult to read and maintain.

Reply:
Thanks for that comment, @Bronek.  I also think that toward zero and away from zero feel a lot more natural.  However I know that the values being operated on are almost always positive, so maybe it's not a big deal.

Reply:
We have `Number::towards_zero`, but not `Number::away_zero`.  If we decide we need `Number::away_zero` I think I could add it fairly quickly.

Reply:
> Regarding the constructor, here's the best I could find. In the C++23 Standard it's section 6.7 Temporary objects [class.temporary] paragraph 2:
> 
> > The materialization of a temporary object is generally delayed as long as possible in order to avoid creating unnecessary temporary objects.
> 
> Note that it says "generally", which I take to mean it is not guaranteed.

@scottschurr doesn't this [sequenced before](https://en.cppreference.com/w/cpp/language/eval_order#:~:text=In%20a%20function%2Dcall%20expression%2C%20the%20expression%20that%20names%20the%20function%20is%20sequenced) rule imply that in `downward()(a + b)`, `downward()` is going to be evaluated first?

Reply:
Yes, I'm pretty sure that in your example `downward()` will be evaluated first.

However I'm most concerned about @HowardHinnant's example at the top of this discussion.  If both `upward()` and `downward()` are constructed before either of the functors are called then one of the computations will round the wrong way.  The only way I personally know of to guarantee order of construction is with a sequence point.

The order of evaluation discussion says very little about when _constructors_ will be called.  My experience with undefined behavior has made me leery of surprises that optimizers can bring.  But I haven't found wording that _proves_ that the compiler will do the wrong thing.

Reply:
I give up (for now)!

Reply:
It seems like the direction we're going is that `upward()` and `downward()` will be a lot less important.  That's probably for the best.  However I cobbled up an approach that is not as pretty as the original `upward()` and `downward()` but includes the necessary sequence points.  It leans heavily on lambdas.  The code looks like this:
```
template<Number::rounding_mode Rm, typename Functor>
requires std::is_invocable_v<Functor>
auto
numberRounder(Functor&& f)
{
    saveNumberRoundMode guard(Number::getround());
    Number::setround(Rm);
    return std::forward<Functor>(f)();
}

template<typename Functor>
requires std::is_invocable_v<Functor>
auto
roundUp(Functor&& f)
{
    return numberRounder<Number::upward>(std::forward<Functor>(f));
}

template<typename Functor>
requires std::is_invocable_v<Functor>
auto
roundDn(Functor&& f)
{
    return numberRounder<Number::downward>(std::forward<Functor>(f));
}
```
Using it looks like this (including clang-formatting):
```
    auto const nTakerPaysConstraint = roundDn([&] {
        return pool.out * targetQuality.rate() -
            roundUp([&] { return pool.in / f; });
    });
```
Just for your consideration.

Also, I've found the two-phase use of `saveNumberRoundMode` to be a bit more awkward than necessary.  So a while back I cobbled up something I find easier to use.  The implementation is in STAmount.cpp, since that's the only place it's currently used:
```
class NumberRoundModeGuard
{
    saveNumberRoundMode saved_;

public:
    explicit NumberRoundModeGuard(Number::rounding_mode mode) noexcept
        : saved_{Number::setround(mode)}
    {
    }

    NumberRoundModeGuard(NumberRoundModeGuard const&) = delete;

    NumberRoundModeGuard&
    operator=(NumberRoundModeGuard const&) = delete;
};
```
I personally find `NumberRoundModeGuard` to be easier to think about and use than `saveNumberRoundMode`.  If that's true for you as well, then I encourage you to move `NumberRoundModeGuard` into the Number.h header and use it freely.

Reply:
Thanks for both suggestions! I was considering lambdas but thought it would not be easy to read. But now that I look at it, I think it's a good alternative.

Reply:
@scottschurr Nice!